### PR TITLE
feat: generate kcl schema from json schema

### PR DESCRIPTION
**Pull request by jakezhu9**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

part of https://github.com/kcl-lang/kcl/issues/526

#### 2. What is the scope of this PR (e.g. component or file name):

pkg/tools/gen

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

In this pull request, I add the capability to convert from JSON schema to KCL schema. Here are the details:

- Expanded the GenKcl function to support various file types and automatic detection of file types.

- Integrated the [jsonschema](https://github.com/qri-io/jsonschema) package for parsing JSON schemas. I made some adjustments to meet our needs, like making some fields public and using ordered maps in the "properties" keyword to preserve the order.

- Added new types, such as Schema, Property, along with the necessary conversion code. I initially considered using the existing `pb.KclType`, but it lacked some fields like validation. So, I opted to create a new type.

- Added template-related functions to generate KCL schema code. I have also added the test code, you can view the results from it.

Please note that some JSON schema keywords are not currently supported, but I plan to add them in the near future. Additionally, I'll be working on documentation and validation sections.

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

File: pkg/tools/gen/types.go [hunk 1]
```diff
@@ -24,6 +24,89 @@ const (
...
+}
+
+type TypeInterface interface {
+	Format() string
+}
+
+type TypePrimitive string
+
+const (
+	Str   TypePrimitive = "str"
```
Comment (Peefy) [inline pkg/tools/gen/types.go]: dup definitions at types.go line 14-24

File: pkg/tools/gen/types.go [hunk 2]
```diff
@@ -24,6 +24,89 @@ const (
 	typNumberMultiplier = "number_multiplier"
 )
 
+type KclSchema struct {
```
Comment (Peefy) [inline pkg/tools/gen/types.go]: Would U pls add more documents?

**Comment by Peefy** [comment]
Good Job! I will carefully review it later, and before that, there are some reminders:
+ Would you pls add more comments and tests for exposed structures and functions?
+ It would be even better if the PR could be split smaller PRs?
+ We recommend U can write your contribution process and ideas as an blog and sharing your labor achievements with the community and users. ❤️

File: pkg/tools/gen/jsonschema/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# jsonschema
```
Comment (Peefy) [inline pkg/tools/gen/jsonschema/README.md]: It is best to place the jsonschema package separately in `third_party` folder at the root folder.

File: pkg/tools/gen/genkcl_jsonschema.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
+package gen
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/iancoleman/strcase"
+	"io"
```
Comment (Peefy) [inline pkg/tools/gen/genkcl_jsonschema.go:6]: U can sort the import statements.

File: pkg/tools/gen/types.go (hunk 1 above)
Comment (jakezhu9) [reply pkg/tools/gen/types.go]: You are right, I will remove these duplicate codes

File: pkg/tools/gen/types.go (hunk 2 above)
Comment (jakezhu9) [reply pkg/tools/gen/types.go]: 👌

File: pkg/tools/gen/jsonschema/README.md (hunk 3 above)
Comment (jakezhu9) [reply pkg/tools/gen/jsonschema/README.md]: Yes! I'll place it to third_party folder as suggested

File: pkg/tools/gen/genkcl_jsonschema.go (hunk 4 above)
Comment (jakezhu9) [reply pkg/tools/gen/genkcl_jsonschema.go:6]: Done

**Comment by jakezhu9** [comment]
Thanks for your careful review! Your feedback really helped improve my code.

> Would you pls add more comments and tests for exposed structures and functions?

Certainly! It also remind me that as a package we should not make our internal type and functions public. I think we should only expose the structures and functions related to generating code, making it easier to use the package.

> It would be even better if the PR could be split smaller PRs?

You are right! Small PRs are easier to understand and review. However, in this PR, most of the changes are in the jsonschema package, and it only implements basic conversion. So, I don't think it needs to be split further. But, I'll definitely consider making smaller PRs in the future.

> We recommend U can write your contribution process and ideas as an blog and sharing your labor achievements with the community and users. ❤️

Good idea! I will try to record this process and share it later

For the kcl type in my code, I designed a typeInterface to support different types, reducing unnecessary fields in each type and making it simpler and clearer. I'm also working on making it more generic so that it can support more features in the future, even though it might seem a bit duplicative with the existing "pb.KclType."

As for the CI error, it's caused by a lower version of Go. If you run it locally with a version above 1.18, it should pass.

Your suggestions have been really inspiring, and if you have any more ideas, please feel free to share them with me!

**Comment by Peefy** [comment]
Good Job! U can rebase the main branch to upgrade the KCL type API and go 1.19+ in CI.

**Pull request merged**